pkg/sensors/accelerator/gpu/source/gpu_dcgm.go

```diff
@@ -239,77 +239,71 @@ func (d *GPUDcgm) GetProcessResourceUtilizationPerDevice(device interface{}, gpu
-		if isMIG { // this is a MIG, get it entity id and reads the related fields
-			for _, val := range vals {
-				if val.FieldId == ratioField {
-					migUtilization := ToFloat64(val, 100)
-					// ratio of active multiprocessors to total multiprocessors
-					// the MIG metrics represent the utilization of the MIG device. We need to normalize the metric to represent the overall GPU utilization
-					// FIXME: the MIG device could have multiple processes, how to split the MIG utilization between the processes?
-					normalizedComputeUtil := migUtilization * multiprocessorCountRatio
-					klog.V(debugLevel).Infof("pid %d computeUtil %f multiprocessor count ratio %v\n", p.Pid, normalizedComputeUtil, multiprocessorCountRatio)
-					processAcceleratorMetrics[p.Pid] = ProcessUtilizationSample{
-						Pid:         p.Pid,
-						TimeStamp:   uint64(time.Now().UnixNano()),
-						ComputeUtil: uint32(normalizedComputeUtil),
+		isMIG := p.GpuInstanceId != 0 && p.GpuInstanceId < maxMIGProfiles // if this is not a MIG, GpuInstanceId is set to e.g. 4294967295
+		if isMIG {                                                        // this is a MIG, get it entity id and reads the related fields
+			// get the MIG info from d.entities
+			for _, migInfo := range d.entities[gpuID] {
+				if p.GpuInstanceId == uint32(migInfo.Info.NvmlInstanceId) {
+					klog.V(debugLevel).Infof("MIG Device %v InstanceId %v GpuIndex %v \n",
+						migInfo.Entity.EntityId, migInfo.Info.NvmlInstanceId, migInfo.Info.NvmlGpuIndex)
+					vals, err = dcgm.EntityGetLatestValues(dcgm.FE_GPU_I, migInfo.Entity.EntityId, deviceFields)
+					if err != nil {
+						klog.V(debugLevel).Infof("failed to get latest values for fields: %v", err)
+						return processAcceleratorMetrics, err
 					}
+					multiprocessorCountRatio = float64(migInfo.Profile.MultiprocessorCount) / migInfo.TotalMultiprocessorCount
+					break
 				}
 			}
-			klog.V(debugLevel).Infof("\n")
-		} else {
-			gpuUtilization := float64(0)
-			if err == nil {
-				for _, val := range vals {
-					if val.FieldId == ratioField {
-						gpuUtilization = ToFloat64(val, 100)
-					}
+		} else { // this is a GPU, get the GPU utilization
+			multiprocessorCountRatio = 1.0
+			vals, err = dcgm.GetLatestValuesForFields(uint(gpuID), deviceFields)
+		}
+		for _, val := range vals {
+			if val.FieldId == ratioField {
+				deviceUtilization := ToFloat64(val, 100)
+				if deviceUtilization <= 0 {
+					deviceUtilization = 0.5 // avoid division by zero
 				}
+				if deviceUtilization > 100 {
+					deviceUtilization = 100
+				}
+				// ratio of active multiprocessors to total multiprocessors
+				// the MIG metrics represent the utilization of the MIG device. We need to normalize the metric to represent the overall GPU utilization
+				// FIXME: the MIG device could have multiple processes, how to split the MIG utilization between the processes?
+				normalizedComputeUtil := deviceUtilization * multiprocessorCountRatio
```

Comment:
What metrics are we targeting over here? If the metrics are related to SM related usage rate (e.g., 0-1), this normalization should be fine. If we also need to query memory related metrics, the normalization rate should be consider as the proportion of MIG memory size to total gpu memory size.

Reply:
At the moment, only SM usage.